### PR TITLE
kubectl apply: warning that kubectl will ignores no-namespaced resource in future release with namespace specified and with default pruneAllowlist

### LINKED ISSUE #110905: kubectl apply: prune should ignore no-namespaced resource if -n is not empty
seems I have  the same problem. save this to a file named `a.yaml`
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    environment: sandbox
    provider: kustomize
  labels:
    app: nginx
  name: nginx
  namespace: sandbox
spec:
  ports:
  - port: 80
  selector:
    app: nginx
  type: NodePort
```

And run `kubectl apply --prune --dry-run=client --all -n sandbox -f a.yaml`, the results
```
service/nginx created (dry run)
namespace/cattle-system pruned (dry run)
```

I have no idea why it want to prune the `namespace/cattle-system`.

_Originally posted by @wd in https://github.com/kubernetes/kubernetes/issues/66430#issuecomment-708922169_

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
It is weird that a non-namespaced resource is pruned when I run `kubectl apply` on a specified namespace.
After discussions in sig-cli meeting, we only decided to add a warning in the current kubectl.
In a future release, we will change the behavior. (at least 1.28 or later)

#### Which issue(s) this PR fixes:
ref #110905

#### Special notes for your reviewer:

https://github.com/kubernetes/kubernetes/blob/fa16bf8e121e9a9ce0a8b92d96a39c986152c484/staging/src/k8s.io/kubectl/pkg/util/prune/prune.go#L41-L58

#### Does this PR introduce a user-facing change?

```release-note
kubectl apply: warning that kubectl will ignore no-namespaced resource `pv & namespace` in a future release if the namespace is specified and allowlist is not specified
```

- [ ] TODO: change the behavior as the comments in a future release: maybe v1.28.

### CODE REVIEW

**Comment by pacoxu** [comment]
/assign @brianpursley @eddiezane 
do you have time to take a look？
No behavior change.

**Comment by brianpursley** [comment]
Prune cane be kind of tricky.

Apply, with `-n` but without `--prune`, does create resources, but ignores the namespace:
```
$ kubectl apply -n foo -f prunetest/
clusterrole.rbac.authorization.k8s.io/secret-reader created
```

I don't know how `--prune` should work in this case.  
* On one hand, it seems like it should work the same way as apply without `--prune`.  
* But on the other hand, it can be surprising if your prune unintentionally matches a label on a non-namespaced resource.

I am inclined to approve this, even though it is inconsistent with how apply without prune works, but I want to see what @seans3 thinks first.

**Comment by KnVerey** [comment]
Another thing to keep in mind is that this issue/PR is framed around the default allow list, which the user can override with `--prune-whitelist=`. In my experience, the defaults are _terrible_ (including for the unpredictable scoping issue reflected here!), so it's best to always override with exactly what you want. For example, [Krane](https://github.com/Shopify/krane) uses pruning under the hood but builds its own allowlist that is restricted to either namespaced or non-namespaced only and that includes CR types (slow and tedious, but behaves predictably).

Point being, we should consider the case where the user is explicitly overriding the list. As @brianpursley rightly points out, a single `apply` can target namespaced and non-namespaced resources at the same time, even with the `--namespace` flag. So if the user does `kubectl apply --prune -n test --prune-whitelist= v1/PersistentVolume,my.co/MyGlobalKind`, I don't think it would be a good experience for the PVs and MyGlobalKinds to be silently skipped.

Maybe another option could be to change the _default_ allow list when the namespace flag is set, but continue to honor any user-specified list. Wdyt?

I'd also be curious to get other SIG members' perspective on whether we need a notice period for a behaviour change like this, e.g. emit a warning for N releases before making the change. It's true that pruning is marked as alpha in the docs, but that status isn't explicit in the actual command/flag, and this behaviour has been around a VERY long time.  So long that someone somewhere likely relies on it, sensibly or not.

**Comment by pacoxu** [comment]
> Point being, we should consider the case where the user is explicitly overriding the list. As @brianpursley rightly points out, a single `apply` can target namespaced and non-namespaced resources at the same time, even with the `--namespace` flag. So if the user does `kubectl apply --prune -n test --prune-whitelist= v1/PersistentVolume,my.co/MyGlobalKind`, I don't think it would be a good experience for the PVs and MyGlobalKinds to be silently skipped.
> 
> Maybe another option could be to change the _default_ allow list when the namespace flag is set, but continue to honor any user-specified list. Wdyt?

Good Point. The current fix is a behavior change and is not consistent with `apply` without `--prune`. 

Two key problems here.
1. the default whitelist for prune is including so much
2. if a user specifies the kind, it is odd to not cover them

The current fix seems to be easier to understand, but cannot cover the above two problems gracefully.


> I'd also be curious to get other SIG members' perspective on whether we need a notice period for a behaviour change like this, e.g. emit a warning for N releases before making the change. It's true that pruning is marked as alpha in the docs, but that status isn't explicit in the actual command/flag, and this behaviour has been around a VERY long time. So long that someone somewhere likely relies on it, sensibly or not.

We should decide which is the final behavior first. Should we ignore the non-namespaced resources finally. Or just warn the user if the flags are not that proper.
A proposal:
**In this v1.26(for instance)**
1. `apply -n xxx` without `--prune`: if there are some non-namespaced resources, give it a warning message.
2. `apply -n xxx` with `--prune`: give a warning if there is non-namespaced resources in the allowlist.
**In this v1.28+(for instance)**
1. `apply -n xxx` without `--prune`: if there are some non-namespaced resources, give it a warning message. (Keep the behavior)
2. `apply -n xxx` with `--prune`: skip non-namespaced resources in the allowlist.

**Comment by KnVerey** [comment]
As much as I personally don't like the idea of mixing scopes in a single apply operation, from an ecosystem perspective, I think that we should not change or warn based on anything the user is doing explicitly. So if they do `apply -n xxx` without `--prune`, and they pass non-namespaced resources, we should continue to handle them without warning. Same if they explicitly give us a prune allow list with a mix of scopes. The problem that I think is worth addressing is specifically the default allow list for `apply -n xxx --prune`, since that default list is unexpectedly not sensitive to the `-n` argument, such that it has what users might reasonably perceive to be a bug or dangerous side-effect. Counter-proposal:

**In the next release:** `apply -n xxx --prune` emits a warning IFF it is NOT given `--prune-whitelist`. The warning should say something like "Deprecated: kubectl apply will no longer prune non-namespaced resources by default when used with the --namespace flag in a future release. To preserve the current behaviour, list the resources you want to target explicitly in the --prune-whitelist flag." Possibly, we could also emit a tailored warning for every pruned resource that will be skipped in the future (but that can't be the only warning, because users who happen to not prune anything on a given instance of an apply command may still have that same command affected in the future).

**After deprecation period passes:** Change the default allow list for `apply -n xxx --prune` to depend on whether `--namespace` is set. If it is set, the allow list should only include namespaced resources. If `--prune-whitelist` is set, there should be no behaviour change.

I personally would support that, but based on https://github.com/kubernetes/kubernetes/pull/109705#issuecomment-1122422743 I suspect Sean may not. No matter what, this PR is proposing a behaviour change to a long-standing command, so a compelling argument for it needs to be brought to the SIG for discussion.

---
Incidentally, regardless of this decision, IMO we should add an alias for this flag that uses inclusive language, i.e. `--prune-allowlist`. It would be best to do that ahead of other changes we make (if any) so that the deprecation message calls out the new name.

**Comment by pacoxu** [comment]
> **In the next release:** `apply -n xxx --prune` emits a warning IFF it is NOT given `--prune-whitelist`. The warning should say something like "Deprecated: kubectl apply will no longer prune non-namespaced resources by default when used with the --namespace flag in a future release. To preserve the current behaviour, list the resources you want to target explicitly in the --prune-whitelist flag." Possibly, we could also emit a tailored warning for every pruned resource that will be skipped in the future (but that can't be the only warning, because users who happen to not prune anything on a given instance of an apply command may still have that same command affected in the future).
> 
> **After deprecation period passes:** Change the default allow list for `apply -n xxx --prune` to depend on whether `--namespace` is set. If it is set, the allow list should only include namespaced resources. If `--prune-whitelist` is set, there should be no behaviour change.

It sounds better to me.

1. `apply -n xxx --prune`:  default `prune-whitelist` should not include non-namespaced resources. Only `ns` and `pv` will be affected. @seans3 (Removing namespace and pv are something dangerous to users.)
2. `apply --prune` without `-n xx`:  default `prune-whitelist` should include non-namespaced resources.

https://github.com/kubernetes/kubernetes/blob/fa16bf8e121e9a9ce0a8b92d96a39c986152c484/staging/src/k8s.io/kubectl/pkg/util/prune/prune.go#L41-L58

**Comment by pacoxu** [comment]
There is a discussion in the sig-cli meeting about this at https://www.youtube.com/watch?v=ZGKfwWof9js.

**Comment by pacoxu** [comment]
/hold
As this is an api-change, sig will not accept the change.
Re-design may be needed if we want user to have a better experience.

**Comment by KnVerey** [comment]
> As this is an api-change, sig will not accept the change.
> Re-design may be needed if we want users to have a better experience.

In the meeting, we agreed that we *would* accept it, but only after adding a warning and waiting for the deprecation period to pass, as we had been discussing on this PR. Specifically, that agreement is here: https://youtu.be/ZGKfwWof9js?t=1750. Are you willing to pursue that process?

**Comment by pacoxu** [comment]
Let me update the pr next week. Thanks for the help and raising the problem in the sig meeting.

**Comment by pacoxu** [comment]
Updated. Add a warning message and a todo comment in code.

**Comment by pacoxu** [comment]
The warning is added. Should we backport it to v1.25~v1.23? So this warning will make all kubectl users to take care of this.

File: staging/src/k8s.io/kubectl/pkg/cmd/diff/prune.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/diff/prune.go:55]: Why is this hardcoded to false?

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/util/prune/prune.go:28]:
```suggestion
// default allowlist of namespaced resources
```

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/util/prune/prune.go:29]: Suggestion: make this `defaultNamespacedPruneResources` and add `defaultNonNamespacedPruneResources` with the other two instead of hardcoding them inline.

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/util/prune/prune.go:57]:
```suggestion
// if namespace is explicitly specified, the default allow list should not include non-namespaced resources.
```

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/util/prune/prune.go:62]: @soltysh @brianpursley for kubectl changes like this, should we be specifying a _specific_ future release here and/or in the user-facing warning?

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/util/prune/prune.go:64]: Don't we want to print this when namespace IS specified, because that's the case where the behaviour will change? Can we test for when this is/isn't printed please?

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go:49]: This seems to be missing coverage for the case where an allow list was explicitly provided. Can we add one, or am I missing that that's covered elsewhere? Ideally I think we'd also have tests for manual lists + namespace provided or not, noting that once we change the behaviour for the default list, it should NOT affect the behaviour for custom lists.

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go:71]: Let's add a note here that this should change two zero after the deprecation period has passed (maybe link to this PR?)

**Review by KnVerey — COMMENTED** [review]
Re: backporting, maybe someone else can correct me, but IIRC the policy is pretty conservative and I don't think error messages are generally eligible.

File: staging/src/k8s.io/kubectl/pkg/cmd/diff/prune.go
Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/cmd/diff/prune.go:55]: `pruneAll()` will get both namespaced and nonNamespaced rest mapping. So here, namespaceSpecified should always be false here.

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune.go
Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/util/prune/prune.go:64]: I make it wrongly.

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go
Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go:71]: Done

File: staging/src/k8s.io/kubectl/pkg/cmd/diff/prune.go
Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/cmd/diff/prune.go:55]: Fixed.

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go
Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go:49]: I added one. 
But from code-coverage aspect, it does not cover more as the default resource is not empty.

```
ok      k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/util/prune  0.475s  coverage: 84.4% of statements
```

**Comment by pacoxu** [comment]
/hold cancel
as there is an agreement in sig-meeting.

**Review by brianpursley — CHANGES_REQUESTED** [review]

**Comment by brianpursley** [comment]
> Incidentally, regardless of this decision, IMO we should add an alias for this flag that uses inclusive language, i.e. `--prune-allowlist`. It would be best to do that ahead of other changes we make (if any) so that the deprecation message calls out the new name.

@KnVerey Yeah, we should go ahead and do that first, that way we won't have to come back and change this warning later.

I made a PR switching to the `--prune-allowlist` flag.

It looks like it assigned you as a reviewer.  Can you take a look?

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune.go
Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/util/prune/prune.go]: Yes, we may change it in 1.29 or later. Fixed.

**Comment by KnVerey** [comment]
Can you please:
* Rebase now that https://github.com/kubernetes/kubernetes/pull/113116 merged
* Update the deprecation warning to use the new flag name
* Add tests covering the default lists at the command level, building on the suite that Brian introduced here: https://github.com/kubernetes/kubernetes/pull/113116/files#diff-73932b6b01149414cf1793cb2742fb9e07fd8a1de292b49852830273e5d30f96

**Comment by pacoxu** [comment]
> Can you please:
> 
> * Rebase now that [kubectl apply: Deprecate --prune-whitelist in favor of --prune-allowlist #113116](https://github.com/kubernetes/kubernetes/pull/113116) merged
> * Update the deprecation warning to use the new flag name
> * Add tests covering the default lists at the command level, building on the suite that Brian introduced here: https://github.com/kubernetes/kubernetes/pull/113116/files#diff-73932b6b01149414cf1793cb2742fb9e07fd8a1de292b49852830273e5d30f96

Updated accordingly.
The main change is in last commit: https://github.com/kubernetes/kubernetes/pull/110907/commits/2f8b0b2899762f97f2725f87dc0fd8a7d1f62afc

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:764]: Won't the test for the case where the namespace flag is NOT provided still need this? If so, I don't think we need the deprecation message.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:793]: removed by mistake?

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:817]: Doesn't this note belong on the second test, i.e. isn't that the case that we'll need to update later?

File: staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go:70]: why would the number of namespaced resources decrease?

Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go]: It is my fault. 
In the future,
```
			expectedns:  14,
			expectednns: 0,
```
It should be like above.

Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/util/prune/prune_test.go]: Updated.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go
Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:793]: This is a copy-paste typo of https://github.com/kubernetes/kubernetes/pull/113116/files#diff-73932b6b01149414cf1793cb2742fb9e07fd8a1de292b49852830273e5d30f96.

The last resource is not applied and no new err is set after the err is already checked.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go
Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go]: Yes, Updated.

It should be with a test case with namespace specified but without pruneAllowlist.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go
Comment (pacoxu) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go]: Removed.

**Comment by brianpursley** [comment]
Latest changes look good.  A few things about the PR itself:
* The title probably needs to be updated to reflect that this is just a warning.
* It also probably shouldn't say it "Fixes #110905" because that will close the issue. It is related to that but doesn't fix it yet.
* Can you squash your commits
I will defer to @KnVerey in case she has other comments

**Review by brianpursley — APPROVED** [review]

**Comment by pacoxu** [comment]
> Latest changes look good. A few things about the PR itself:
> 
> * The title probably needs to be updated to reflect that this is just a warning.
> * It also probably shouldn't say it "Fixes # 110905" because that will close the issue. It is related to that but doesn't fix it yet.
> * Can you squash your commits
>   I will defer to @KnVerey in case she has other comments

Updated and also the release note to use allowlist instead of whitelist.

**Comment by collimarco** [comment]
This change is confusing and the warning is totally obscure.

For example I use Cuber to compile and apply this YML file:
https://github.com/cuber-cloud/cuber-gem/blob/master/lib/cuber/templates/deployment.yml.erb

That YML file contains everything that goes into the namespace.

Then I `apply` it to Kubernetes using `--prune`:
https://github.com/cuber-cloud/cuber-gem/blob/5f4a708bc511c44514129185226db139807cdd8d/lib/cuber/commands/deploy.rb#L86

Why do I get this warning now? 

```
W0723 20:57:03.499655    2113 prune.go:71] Deprecated: kubectl apply will no longer prune non-namespaced resources by default when used with the --namespace flag in a future release. To preserve the current behaviour, list the resources you want to target explicitly in the --prune-allowlist flag.
```

**What should I do? I simply want to apply that YML file to the namespace and prune every resource in the namespace that is no longer present in the YML file.**

**Comment by pacoxu** [comment]
@collimarco in your provide yaml, the only non-namespaced resource is the namespace itself.

I think this warning will not affect your application.

You have specified the namespace here, and even if you remove the namespace object in the yaml, the `apply --prune -n app` will not remove the namespace resource in the future version. 
- In the current and old versions, if you remove `the namespace part` in your yaml, and run the `apply --prune -n app`, then the app namespace will be removed. (This is very dangerous. We just want to avoid it happening in the future.) 

You can see an example in https://github.com/kubernetes/kubernetes/issues/110905.

**Comment by collimarco** [comment]
@pacoxu Perfect, thanks for the clarification. Is there any way to get rid of the warning message? Or will it disappear automatically in the future releases of kubectl?

**Comment by pacoxu** [comment]
> Is there any way to get rid of the warning message? Or will it disappear automatically in the future releases of kubectl?

We plan to remove the warning in v1.29, and I think the behavior will change in v1.29. But it is still a plan and it should be done with the agreement of SIG-Cli maintainers.

**Pull request merged**
